src/editor/model.js

```diff
@@ -90,6 +90,11 @@ export default class EditorModel {
         }
     }
 
+    reset(serializedParts, caret, inputType) {
+        this._parts = serializedParts.map(p => this._partCreator.deserializePart(p));
+        this._updateCallback(caret, inputType);
+    }
+
     update(newValue, inputType, caret) {
         const diff = this._diff(newValue, inputType, caret);
```

Comment:
So we're diffing on every keystroke I guess? How expensive is that? Does this processing make the input experience feel sluggish at all? (I assume it shouldn't be sluggish at least, since the browser has already displayed the new character before the handler runs...?)

Reply:
why does this need a diffing algorithm rather than just snapshotting the editor state on every word boundary?

Reply:
For clarity, the diffing is part of how the editor works, and is unrelated to the undo support.

The diffing algorithm is very basic, and assumes there is only one change around the caret position, so this should be very inexpensive. See `editor/diff.js` for details. The most expensive part is probably to turn the content-editable content into a single string (with some logic, see `getCaretOffsetAndText`)  on every keystroke. Some preliminary profiling shows that we can probably make some improvements still, and there is an item on https://github.com/vector-im/riot-web/issues/9866 for that.

As for the reason to use diffing; the alternative would be to recreate the model on every keystroke based on the DOM of the editor. Without considering the runtime cost/benefits of this, it would not allow us preserve state for a given section of the editor, like the part type, or the userid for a pill.

So on every keystroke, the editor dom is turned into a string, and the model updated: the model turns its current state into text, and compares the current and new text value. The difference between both is then applied to the existing parts of the model state, applying validation logic in the parts.

The benefit of this is that we can use the `input` event, which is broadly supported, to find changes in the editor. We don't have to rely on keyboard events, which poorly relate to text input or changes.

Once the parts of the model are updated, the DOM of the editor is then reconciled with the new model state, see `renderModel` in `editor/render.js` for this. If the model didn't reject the input and didn't make any additional changes, this won't make any changes to the DOM at all, and should thus be fairly efficient.

I should probably document all of this somewhere.

Reply:
ah, ok. i would *definitely* put all this into a massive /**/ comment at the top of the editor file, or possibly even docs/brunoEditor.md or whatever we're calling this ;)

Reply:
https://github.com/matrix-org/matrix-react-sdk/pull/3272 adds some documentation.